Introducing the OemrUI Class - Take 3

zbig01 opened the pull request:
Created a single associative array as suggested
```
$arrOeUiSettings = array(
    'heading_title' => xl('The page Heading'),
    'include_patient_name' => false,
    'expandable' => false,
    'expandable_files' => array(),//all file names need suffix _xpd
    'action' => "",//conceal, reveal, search, reset, link or back
    'action_title' => "",
    'action_href' => "",//only for actions - reset, link or back
    'show_help_icon' => false,
    'help_file_name' => ""
);

```
now all elements are echoed out as functions

```
<div id="container_div" class="<?php echo $oemr_ui->oeContainer();?>">

<?php echo  $oemr_ui->pageHeading() . "\r\n"; ?>

<?php $oemr_ui->helpFileModal();?>
```

Refer to PR #2016, #1991 for previous comments

bradymiller commented on composer.json:
```diff
@@ -36,9 +36,7 @@
         "zendframework/zendframework" : "2.5.3",
-        "particle/validator": "2.3.3",
-        "guzzlehttp/guzzle": "6.3.3",
-        "kamermans/guzzle-oauth2-subscriber": "1.0.5"
```
add back above changes

bradymiller commented on composer.json:
```diff
@@ -87,7 +85,7 @@
             "OpenEMR\\Pdf\\" : "library/pdf/src",
             "OpenEMR\\Tabs\\" : "library/tabs/src",
             "OpenEMR\\Rx\\Weno\\" : "library/weno/src",
-            "HL7\\FHIR\\STU3\\": "phpfhir/HL7/FHIR/STU3/src"
```
add back above line (ensure a comma is added to end

bradymiller commented on interface/billing/sl_eob_search.php:
```diff
@@ -675,43 +676,41 @@ function toEncSummary(pid) {
-    require_once("$srcdir/expand_contract_inc.php");
+    if ($arrOeUiSettings['include_patient_name']) {
+        $arrOeUiSettings['heading_title'] .= " - " . getPatientNameFirstLast($pid);
+    }
```
oh, this is looking nice. One last thing to really clean up this class is to migrate the above 6 lines into the constructor of the OemrUI class. (then can say bye bye to the DO NOT EDIT BELOW section)

bradymiller replied:
I'll look into this issue. We should be able to make it work :)

bradymiller replied:
The getPatientNameFirstLast and collectAndOrganizeExpandSetting should be able to work within a class (ie. I did a quick look at them to ensure they were not trying to call and out of scope variables). How did you try to call them within the class. Note $pid won't work within the class and will need to use $_SESSION['pid']

zbig01 replied:
I don't think we can get regular functions to work inside a class. If my understanding of a class being a template used to create an object is correct these functions won't work inside the template.

I tried to place it in the constructor but could not get it to work :(

zbig01 commented on library/js/oeUI/universalTooltip.js:
```diff
@@ -0,0 +1,28 @@
+ * @copyright Copyright (c) 2018 Ranganath Pathak <pathak@scrs1.org>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+?>
```
The universalTooltip.js needs to be included whenever the class OemrUI  is called. Can you think of a convenient place where it can be placed without having to remember yet another script to include

bradymiller replied:
Would it make sense to just always include it? (ie. no matter what like the default scripts are in the Header function call)

zbig01 replied:
To function it need jquery-ui, not always included in pages. Unless you want to modify the `openemr/config/config.yaml` file

bradymiller replied:
Per the class, it looks like both jquery-ui and jquery-ui-base are also required. Is this true? Then will think about best way to easily support this in these scripts.

zbig01 replied:
yes the base is also needed. I tried to find a place to put it but could not think of a suitable one, hence the ugly hack

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            $help_icon = $this->helpIcon();
+            $expandable_icon = $arrexpandIcon[0];
+            $container = $arrexpandIcon[1];
+            $heading = "<h2>$heading $expandable_icon $action_icon $help_icon</h2>";
```
```
$heading = "<h2>" . text($heading) . " $expandable_icon $action_icon $help_icon</h2>";
```

zbig01 replied:
already done :)
$heading = text($this->heading);

bradymiller replied:
lol, missed that

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            $heading = "<h2>$heading $expandable_icon $action_icon $help_icon</h2>";
+        } else {
+            $heading = "<h2>" . xlt("Please supply a heading") . " <i class='fa fa-smile-o' aria-hidden='true'></i></h2>";
+            $container = 'container';
```
is this variable being used here?

bradymiller replied:
I think above line can be removed

bradymiller commented on interface/billing/sl_eob_search.php:
```diff
@@ -675,43 +676,41 @@ function toEncSummary(pid) {
+    'expandable' => 1,
+    'expandable_files' => array('sl_eob_search_xpd'),//all file names need suffix _xpd
+    'action' => "reset",
+    'action_title' => "",
```
btw, when a title is entered here, then it should be translated via xl function (note do not html escape it here though)

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+        $action_href = ($action_href) ? $action_href : "#";
+        switch ($action) {
+            case "reset":
+                $action_title = ($action_title) ? $action_title : "Reset";
```
Translate the Reset with xl() function

zbig01 replied:
Done here :)
`$action_icon = "<a href='" . attr($action_href) ."' onclick='top.restoreSession()'><i id='advanced-action' class='fa fa-undo fa-2x small' title='" . xla($action_title) ."' aria-hidden='true'></i></a>";`

bradymiller replied:
There are generally 2 forces at work on the translation and escaping:
1. For translations, it is much better to translate as upstream as possible, so the constants get captured in the scripts that grab them from the codebase.
2. For html escaping, best to escape as downstream as possible.
Which is why should translate the Reset on the same line (or when it is entered into the class call), but do the escape call when it is outputted to html.

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+        switch ($action) {
+            case "reset":
+                $action_title = ($action_title) ? $action_title : "Reset";
+                $action_icon = "<a href='" . attr($action_href) ."' onclick='top.restoreSession()'><i id='advanced-action' class='fa fa-undo fa-2x small' title='" . xla($action_title) ."' aria-hidden='true'></i></a>";
```
Change the xla() to attr()

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+                break;
+            case "conceal":
+                $action_title = "Click to Hide"; // default needed for jQuery to function
+                $action_icon = "<i id='show_hide' class='fa fa-2x small fa-eye-slash' title='" . xla($action_title) . "'></i>";
```
Change the xla() to attr()

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+                break;
+            case "reveal":
+                $action_title = "Click to Show"; // default needed for jQuery to function
+                $action_icon = "<i id='show_hide' class='fa fa-2x small fa-eye' title='" . xla($action_title) . "'></i>";
```
Change the xla() to attr()

zbig01 replied:
Is it a problem if we do` xla` instead of` xl `and `attr` ?

bradymiller replied:
see my comment I just posted

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+                break;
+            case "search":
+                $action_title = "Click to show search"; // default needed for jQuery to function
+                $action_icon = "<i id='show_hide' class='fa fa-search-plus fa-2x small' title='" . xla($action_title) . "'></i>";
```
Change the xla() to attr()

zbig01 replied:
ditto

bradymiller replied:
see my comment I just posted

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+                    $target = '_self';
+                }
+                $action_title = ($action_title) ? $action_title : "Click to go to page";
+                $action_icon = "<a href='" . attr($action_href) . "' target = '" .attr($target)."' onclick='top.restoreSession()'><i id='advanced-action' class='fa fa-external-link fa-2x small' title='" . xla($action_title) ."' aria-hidden='true'></i></a>";
```
Change the xla() to attr()

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+                } elseif ($_SESSION ['language_direction'] == 'rtl') {
+                    $arrow_direction = 'fa-arrow-circle-right';
+                }
+                $action_icon = "<a href='" . attr($action_href) ."' onclick='top.restoreSession()'><i id='advanced-action' class='fa " . attr($arrow_direction) . " fa-2x small' title='" . xla($action_title) ."' aria-hidden='true'></i></a>";
```
Change the xla() to attr()

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+        $display_help_icon = $this->display_help_icon;
+        if ($display_help_icon) {
+            if ($_SESSION ['language_direction'] == 'ltr') {
+                $help_icon_title = "To enable help - Go to the User Name on top right > Settings > Features > Enable Help Modal";
```
wrap this with xl() function

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            if ($_SESSION ['language_direction'] == 'ltr') {
+                $help_icon_title = "To enable help - Go to the User Name on top right > Settings > Features > Enable Help Modal";
+            } elseif ($_SESSION ['language_direction'] == 'rtl') {
+                $help_icon_title = "To enable help - Go to the User Name on top left > Settings > Features > Enable Help Modal";
```
wrap this with xl() function

zbig01 replied:
I thought we ought to translate and escape as close to the actual html output as possible

bradymiller replied:
see my comment I just posted

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            if ($GLOBALS['enable_help'] == 1) {
+                $help_icon = '<a class="oe-pull-away oe-help-redirect" data-target="#myModal" data-toggle="modal" href="#" id="help-href" name="help-href" style="color:#676666" title="' . xla("Click to view Help") . '"><i class="fa fa-question-circle" aria-hidden="true"></i></a>';
+            } elseif ($GLOBALS['enable_help'] == 2) {
+                $help_icon = '<a class="oe-pull-away oe-help-redirect" data-target="#myModal" data-toggle="modal" href="#" id="help-href" name="help-href" style="color:#DCD6D0 !Important" title="' . xla($help_icon_title) . '"><i class="fa fa-question-circle" aria-hidden="true"></i></a>';
```
Change the xla() to attr()

zbig01 commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+    private $jquery_draggable;
+    private $modal_body;
+    private $print;
+    private $target;
```
I checked, they are being used

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            $modal_body = "<iframe src=\"$help_file\" id='targetiframe' style='height:100%; width:100%; overflow-x: hidden; border:none'
+                                allowtransparency='true'></iframe>";
+        } else {
+            $modal_body = "<h3> <i class='fa fa-exclamation-triangle  oe-text-red' aria-hidden='true'></i> " . xlt("Check if a help file exists for this page in") . " " . text("Documentation/help_files") . ".<br><br>" . xlt("Then pass it's name as an argument in") . " " . text("function helpFileModal"). ".<br><br>" . xlt("If the help file does not exist create one and place it in") . " " . text(" ...
```
$modal_body = "<h3> <i class='fa fa-exclamation-triangle  oe-text-red' aria-hidden='true'></i> " . xlt("Check if a help file exists for this page in Documentation/help_files") . ".<br><br>" . xlt("Then pass it's name as an argument in function helpFileModal"). ".<br><br>" . xlt("If the help file does not exist create one and place it in Documentation/help_files") . ".<br>" . "</h3>";

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+                <div class="modal-dialog modal-lg">
+                    <div class="modal-content  oe-modal-content" style="height:700px">
+                        <div class="modal-header clearfix">
+                            <button type="button" class="close" data-dismiss="modal" aria-label=$close>
```
```
<button type="button" class="close" data-dismiss="modal" aria-label="$close">
```

zbig01 replied:
will do

bradymiller commented on library/js/oeUI/headerTitleAction.js:
```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Expand Contract State and Show /Hide State js
```
place a .php at the end of this or else developers will get confused (for example, they may try to bring it in as a pure js script which will break):
headerTitleAction.js.php

zbig01 replied:
can you explain why the php is getting parsed correctly in the js file ?

bradymiller replied:
I looked into this, and it is because you are including it within a php file as a php file (using the require call). So, this is why the apache/nginx don't get involved. Still, though, this will thoroughly confuse developers.

bradymiller commented:
Code review complete. I'll get an Up For Grabs demo soon.

zbig01 commented:
I will work on the above. Will add php extension to the above js file.

Can we leave it as `xla` instead of `xl  + attr` ?

zbig01 commented:
@bradymiller 
will translate first and then escape :)

zbig01 commented:
Translated early and escaped late in `OemrUI.php `:)
Changed the `headerTitleAction.js` to `headerTitleAction.js.php`
Changed` universalTooltip.js` phpdoc comment to javadoc comment :)
Changed all scripts that referenced ` headerTitleAction.js` to
`headerTitleAction.js.php`

zbig01 commented:
@bradymiller
Could you point the demo to this PR. Thanks 😀

zbig01 commented:
@bradymiller 
I have cleaned up the code now it looks like this

```
<?php
    $arrOeUiSettings = array(
        'heading_title' => xl('Access Control List Administration'),
        'include_patient_name' => false,
        'expandable' => false,
        'expandable_files' => array(),//all file names need suffix _xpd
        'action' => "link",//conceal, reveal, search, reset, link or back
        'action_title' => "",
        'action_href' => "../../gacl/admin/acl_admin.php",//only for actions - reset, link or back
        'show_help_icon' => true,
        'help_file_name' => "adminacl_help.php"
    );
    $oemr_ui = new OemrUI($arrOeUiSettings);
?>
```

I will update the PR shortly :)

bradymiller commented:
Up For Grabs demo of this PR here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Lambda_-_Up_For_Grabs_Demo

zbig01 commented:
@bradymiller 

The deed is done. The code has been slimmed down. Hope it fits the bill

bradymiller commented on interface/patient_file/reminder/patient_reminders.php:
```diff
@@ -67,273 +68,293 @@ function sel_patient() {
-if ($sortby == "") {
-    $sortby = $sort[0];
-}
+$oemr_ui = new OemrUI(); //to display heading with selected icons and help modal if needed
```
delete above

zbig01 replied:
will do

bradymiller commented on interface/patient_file/report/patient_report.php:
```diff
@@ -58,12 +59,48 @@ function show_date_fun(){
+$oemr_ui = new OemrUI($arrHeading);
+$arr_display_heading = $oemr_ui->pageHeading(); // returns an indexed array containing heading string with selected icons and container string value
+$heading = $arr_display_heading[0];
+$container = $arr_display_heading[1];// if you want page to always open as full-width override the default returned value with $container = 'container-fluid'
```
I think above can go?

zbig01 replied:
My bad:)

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            $action_icon = $this->actionIcon();
+            $help_icon = $this->helpIcon();
+            $expandable_icon = $arrexpandIcon[0];
+            $container = $arrexpandIcon[1];
```
I think above line can be removed

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+            $modal_body = "<iframe src=\"$help_file\" id='targetiframe' style='height:100%; width:100%; overflow-x: hidden; border:none'
+                                allowtransparency='true'></iframe>";
+        } else {
+            $modal_body = "<h3> <i class='fa fa-exclamation-triangle  oe-text-red' aria-hidden='true'></i> " . xlt("Check if a help file exists for this page in") . " " . text("Documentation/help_files") . ".<br><br>" . xlt("Then pass it's name as a value to the element" ." " . text("'help_file_name'") . " "  .  "in the associative array") . " " . text("\$arrOeUiSettings"). ".<br><br>" . xlt("If  ...
```
Below will make much more sense for translators:
```
$modal_body = "<h3> <i class='fa fa-exclamation-triangle  oe-text-red' aria-hidden='true'></i> " . xlt("Check if a help file exists for this page in Documentation/help_files.") . "<br><br>" . xlt("Then pass it's name as a value to the element 'help_file_name' in the associative array \$arrOeUiSettings."). "<br><br>" . xlt("If the help file does not exist create one and place it in Documentation/help_files.") . "<br>" . "</h3>";
```

zbig01 replied:
How would we prevent the following strings `'help_file_name'` ,` $arrOeUiSettings` and` Documentation/help_files` from being translated? 

They would need to be represented as string literals

bradymiller replied:
Could do this
```
$modal_body = "<h3> <i class='fa fa-exclamation-triangle  oe-text-red' aria-hidden='true'></i> " . xlt("Check if a help file exists for this page in Documentation/help_files.{{do not translate Documentation/help_files}}") . "<br><br>" . xlt("Then pass it's name as a value to the element 'help_file_name' in the associative array \$arrOeUiSettings.{{do not translate arrOeUiSettings}}"). "<br><br>" . xlt("If the help file does not exist create one and place it in Documentation/help_files.{{do not translate Documentation/help_files}}") . "<br>" . "</h3>";
```

bradymiller replied:
Those comment in {{ }} are shown to translators but do not show when displayed.

bradymiller commented on library/oeUI/src/OemrUI.php:
```diff
@@ -0,0 +1,290 @@
+        if ($arrOeUiSettings['expandable'] && $arrOeUiSettings['expandable_files']) {
+            $this->current_state = collectAndOrganizeExpandSetting($arrOeUiSettings['expandable_files']);
+        }
+        echo "\r\n<script src='" . $GLOBALS['webroot'] . "/library/js/oeUI/universalTooltip.js?v=" . $v_js_includes. "'></script>\r\n";
```
There are still 2 odd things going on that am trying to figure out :)
1. It's really odd to have constructor echo something like this(forcing folks to initiate the class in the right spot).
2. It's really odd to bring in assets (jquery-ui and the base theme) when initiate the class.(there will be clashing since developers will never know to look in the class; also bad things could happen if they want to bring in another jquery-ui theme rather than base)

Would this solution make sense:
1. Make it a requirement for scripts that use this class bring in via Header function the jquery-ui and a jquery ui theme (at least base). This will make sense to developers and will avoid clashing.
2. Then we can rig config.yml script to always bring in universalTooltip.js if jquery-ui is brought in

zbig01 replied:
I had structured it in this way to minimize adding extra code to get the class to work. As I had mentioned before `jquery-ui `was not being used in very many pages. We can certainly make it an autoloadable  feature in `config.yml`. I have not delved deeply into `jquery-ui ` but I did note that the `tooltip ` function of `bootstrap `and `jquery-ui ` differ in syntax. If `jquery-ui ` is included code using the` bootstrap tooltip ` syntax will fail. 

With regards initiating the class in the right spot I feel that placing the 

```
<?php
    $arrOeUiSettings = array(
        'heading_title' => xl('Access Control List Administration'),
        'include_patient_name' => false,
        'expandable' => false,
        'expandable_files' => array(),//all file names need suffix _xpd
        'action' => "link",//conceal, reveal, search, reset, link or back
        'action_title' => "",
        'action_href' => "../../gacl/admin/acl_admin.php",//only for actions - reset, link or back
        'show_help_icon' => true,
        'help_file_name' => "adminacl_help.php"
    );
    $oemr_ui = new OemrUI($arrOeUiSettings);
?>
```
code block as the last item before the `</head>` tag is intuitive and logical and thus the js that is being echoed will be in the `<head></head>` section of the page.

Furthermore I think it would be useful if we make it a convention to mark the end of the container div with 
`</div><!--end of container div-->`
and place 
`<?php $oemr_ui->helpFileModal();?>`
`<script> <?php require_once("$srcdir/js/oeUI/headerTitleAction.js"); ?></script>`
as the next two lines of code in order to 'standardize' the way the page source code will be rendered

bradymiller replied:
Hi,
You've convinced me :)
I like the end of container div idea.
Regarding the headerTitleAction.js, I noted some scripts use it and some don't; would it make sense to rename the helpFileModal() to something else like oeUIbottom which will do both the helpFileModal and /headerTitleAction.js on the same call (and can decide what to include)?

zbig01 replied:
Let me give it some thought. 
I kind of like `helpFileModal` as it conveys the purpose of the function. I suppose I could add` headerTitleAction.js.php` to the class but I feel leaving it as a js file and have the developer add these two as needed would minimize adding unnecessary code to every page.

bradymiller replied:
This is up to you.
My thoughts are that you are already telling the class if there is a helpfile and if it is expandable. So, might as well let the class decide what to add at with a call to the UiBottom() function, which will always go below the </div><!--end of container div--> thing. This then opens door to adding other functionality in the future without much work.

zbig01 replied:
I always fear the thought of ending up with a monolithic class that contains every possible thing imaginable

bradymiller replied:
lol, that would be awesome :)
It's already a nice class. Decision is yours.

bradymiller commented:
This is really looking clean and getting razor close to getting into the codebase. Still a couple more minor things to work out :)
